### ISSUE TITLE
http1 requests with TE header fail if backend server uses proto h2

### ISSUE THREAD

**Issue by bjacke** [haproxy/haproxy#464]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.0.12-1~bpo10+1 2019/12/21 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-UpVGSu/haproxy-2.0.12=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace

Linux hr2 4.19.0-6-amd64 #1 SMP Debian 4.19.67-2+deb10u2 (2019-11-11) x86_64 GNU/Linux

```

## What's the configuration?

```
# frontend supports HTTP 1.1 and HTTP2:
frontend ft_www
  bind 1.2.3.4:443 tfo alpn h2,http/1.1 accept-proxy ssl crt ...
  use_backend bk_plain

# backend server uses HTTP2:
backend bk_plain
    server bk 10.11.0.3:8080 proto h2

```

## Steps to reproduce the behavior

```
(cat ; sleep 2) <<EOF | openssl s_client -servername www.example.com -connect www.example.com:443
GET / HTTP/1.1
TE: deflate, gzip;q=0.3
Connection: TE, close
Host: www.example.com
User-Agent: Debian uscan

EOF
```

## Actual behavior

haproxy closes the connection immediately. 

## Expected behavior

the client request is valid and haproxy should reply correclty with the backend server's result.

## Do you have any idea what may have caused this?

The reason seems to be the TE http header. The client's TE header for the HTTP 1,1 request is valid but for a HTTP2 request the TE header is only valid if it also contains "trailers", see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/TE

## Do you have an idea how to solve the issue?

haproxy should modify the TE header if the frontend received a HTTP < 2 request from a client and if the backend server is using HTTP2.
If you modify the request either by removing the TE header or by setting only "trailers" in the TE header, then haproxy sends the correct result as this is valid for HTTP2 (as used by the backend server) also.

**Comment by wtarreau** [comment]
Hi!

thanks for your report, it's indeed a stupid bug. It's been addressed with commit bb2c4ae06.

**Comment by wtarreau** [comment]
now backported as far as 1.9, closing.
